File: habit-tracker/src/main/java/com/habit/habit_tracker/security/UserServiceImpl.java
package com.habit.habit_tracker.security;

import java.util.Optional;

import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;

import com.habit.habit_tracker.domain.User;
import com.habit.habit_tracker.repository.UserRepository;

@Service
public class UserServiceImpl implements UserDetailsService {

    private final UserRepository userRepository;

    public UserServiceImpl(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        Optional<User> user = userRepository.findByUsername(username);

        if (user.isEmpty()) {
            throw new UsernameNotFoundException("User not found with username: " + username);
        }

        return UserPrincipal.create(user.get());
    }

    public UserDetails loadUserById(Long userId) throws UsernameNotFoundException {
        Optional<User> user = userRepository.findById(userId);

        if (user.isEmpty()) {
            throw new UsernameNotFoundException("User not found with userId: " + userId);
        }

        return UserPrincipal.create(user.get());
    }
}

File: habit-tracker/src/main/kotlin/com/habit/habit_tracker/security/UserServiceImpl.kt
package com.habit.habit_tracker.security

import com.habit.habit_tracker.domain.User
import com.habit.habit_tracker.repository.UserRepository
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service

@Service
class UserServiceImpl(private val userRepository: UserRepository) : UserDetailsService {

    override fun loadUserByUsername(username: String): UserDetails {
        val user = userRepository.findByUsername(username)
            .orElseThrow { UsernameNotFoundException("User not found with username: $username") }
        return UserPrincipal.create(user)
    }

    fun loadUserById(userId: Long): UserDetails {
        val user = userRepository.findById(userId)
            .orElseThrow { UsernameNotFoundException("User not found with userId: $userId") }
        return UserPrincipal.create(user)
    }
}
